app/post_helpers.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/v6/model"
+)
+
+type postAccessibleBounds struct {
+	accessible   int
+	inaccessible int
+}
+
+func (p postAccessibleBounds) allAccessible() bool {
+	return p.accessible == allAccessiblePosts.accessible && p.inaccessible == allAccessiblePosts.inaccessible
+}
+
+func (p postAccessibleBounds) allInaccessible() bool {
+	return p.accessible == noAccessiblePosts.accessible && p.inaccessible == noAccessiblePosts.inaccessible
+}
+
+var noAccessiblePosts = postAccessibleBounds{accessible: -1, inaccessible: 0}
+var allAccessiblePosts = postAccessibleBounds{accessible: 0, inaccessible: -1}
+
+func getDistance(a, b int) int {
+	if a-b > 0 {
+		return a - b
+	}
+	return b - a
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// getTimeSortedPostAccessibleBounds returns what the boundaries for accessible and inaccessible posts.
+// It assumes that CreateAt time for posts is monotonically increasing or decreasing.
+// It could be either because posts can be returned in ascending or descending time order.
+// Because it returns the boundaries, it is necessary to check whether the accessible or
+// inaccessible post index is greater, and do the filtering accordingly.
+// Special values (which can be checked with methods `allAccessible` and `allInaccessible`)
+// denote if all or none of the posts are accessible
+func getTimeSortedPostAccessibleBounds(postList *model.PostList, earliestAccessibleTime int64) postAccessibleBounds {
```

Comment:
This entire implementation seems too complicated to me. I'll revisit this in the second pass.